### ISSUE TITLE
Cross-tokenizer distillation fails in GKD and MiniLLM trainers

### ISSUE THREAD

**Issue by sambhavnoobcoder** [huggingface/trl#4562]
## Description

When using GKD or MiniLLM trainers for on-policy knowledge distillation with student and teacher models that have **different tokenizers**, the training produces incorrect results. The teacher model receives student-tokenized input without re-tokenization, causing teacher logprobs to be computed in low-probability regions.

## Original Source

This issue was brought to light by [@HeMuyu0327](https://x.com/HeMuyu0327/status/1987382662328168905):
https://x.com/HeMuyu0327/status/1987382662328168905

### What Happens

In the current implementation:
1. Student model generates rollouts using its own tokenizer (e.g., Qwen tokenizer)
2. These student-tokenized IDs are passed **directly** to the teacher model
3. Teacher model (e.g., Llama) interprets these IDs using **its own vocabulary mapping**
4. Since the vocabularies differ, the teacher sees completely different tokens than intended
5. Teacher logprobs are computed on nonsensical sequences → incorrect training signal

### Example

Consider this concrete example:
```
Student (Qwen) tokenizes "Hello" → token ID 123
- In Qwen's vocabulary: ID 123 = "Hello" ✓
- In Llama's vocabulary: ID 123 = "World" ✗

Teacher (Llama) receives ID 123 and computes logprobs for "World" instead of "Hello"
→ Wrong probability distribution!
```

### Minimal Example

```python
from trl.experimental.gkd import GKDConfig, GKDTrainer
from transformers import AutoModelForCausalLM, AutoTokenizer
from datasets import load_dataset

# Load models with DIFFERENT tokenizers
student_model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B")
teacher_model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-3.2-1B")
student_tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B")

dataset = load_dataset("trl-internal-testing/zen", "conversational_language_modeling", split="train")

config = GKDConfig(output_dir="./output", max_steps=10)

trainer = GKDTrainer(
    model=student_model,
    teacher_model=teacher_model,
    args=config,
    train_dataset=dataset,
    processing_class=student_tokenizer,
)

# This will produce incorrect results because:
# - Student generates with Qwen tokenizer
# - Teacher receives Qwen-tokenized IDs
# - Teacher interprets IDs using Llama vocabulary
# - Teacher logprobs computed on wrong tokens!
trainer.train()
```

### Expected Behavior

Training should work correctly with different tokenizers by:
1. Preserving the **text** from student-generated rollouts
2. Re-tokenizing the text using the **teacher's tokenizer**
3. Computing teacher logprobs on correctly tokenized sequences

### Actual Behavior

Training either:
- Produces incorrect results (teacher logprobs computed on wrong tokens)
- Fails with cryptic errors
- "Succeeds" but with degraded distillation quality

## Environment

- **TRL version**: `main` branch (commit: 9bc62066)
- **Transformers version**: Latest
- **Python version**: 3.9+
- **Affected components**: `GKDTrainer`, `MiniLLMTrainer`

**Comment by sambhavnoobcoder** [comment]
I have authored a fix for the same in #4561 . I would really appreciate your reviews on the same , so that we can move fast and get this merged and issue resolved rapidly .

**Comment by kashif** [comment]
So, the issue is inherent to models that distill using some divergence loss, like KL, as it requires the vocabulary size to match between the teacher and student. How did you solve that issue?

**Comment by sambhavnoobcoder** [comment]
correct , so KL divergence typically requires matching vocabulary sizes . however this turns out to be a two fold problem , because primarily there could be a token id mismatch , since different token ids could refer to different tokens in different tokenisers , and alternatively different vocabulary sizes would need special handeling . 

So as a fix , instead of passing student tokenized IDs directly to the teacher, i preserve text from student generated rollouts , which i re-tokenize using the teachers tokenizer and then the teacher can process its own correctly tokenized input. This solves the issue with token id mismatch . 

As for different vocabulary sizes , i use a text-aligned approach where i get the teachers predictions , which i decode to text , then re-encode this using student's tokenizer to convert to the student's vocab space and then use that instead . As for loss calc , i moved from KL to CE loss , since fundamentally KL is reliant on vocab size as we cannot compute KL(P_student || P_teacher) since they live in different dimensional spaces , and that wont change even after the fix . Cross-entropy loss solves this by computing the loss entirely in the student's vocabulary space, making it the correct alternative. 

essentially i am attempting to use text as the bridge between different tokenizer spaces, rather than trying to align vocabularies directly. I am acknowledging the slight tradeoff with the encode / decode overhead , however i am in favour of it due to improvement in correctness since teacher logprobs are now computed on the right tokens.

I hope this answers your question . 

Also , Thank you very much — I didn't expect this to attract attention so quickly from the community. I appreciate your interest in my contribution and would be pleased to answer any questions.

**Comment by kashif** [comment]
right but in GKD and miniLLM the teacher never predicts anything... we have prompts and completions, and for a certain prob. we sample some completions given the prompt from the student model and obtain the detached teacher's logprobs from the student predictions... the loss is then the kl div. loss of the detached teacher logprobs and student logprobs, which requires the teacher/student vocab size to match.

i do not think cross entropy is going to help here.

**Comment by sambhavnoobcoder** [comment]
Thanks @kashif - quick recap: Originally fixed token ID mismatches via text re-tokenization (student text → teacher tokenizer), which solved the cross-tokenizer input issue. But then we used teacher predictions (argmax) decoded/re-encoded as pseudo-labels for CE loss—which as you pointed out, made the teacher generate instead of evaluate, violating GKD's on-policy objective. Now fixed with sequence-level KL: Both models evaluate the same student-generated text (teacher gets re-tokenized version). We compute summed log-probs normalized by token count, with EMA baseline for stability. This keeps it evaluation-only, on-policy, and properly uses KL divergence instead of CE. Tests verify: teacher never calls argmax, loss is KL-based, backward compatibility maintained. Better?

**Comment by kashif** [comment]
@sambhavnoobcoder so the KL div will not work as the teacher token ids do not correspond the student ids when the tokenizers differ and also the KL will not work when the vocab size differ... there is unfortunately no band-aid fix to this issue that you raise and there has been efforts trying to address this. I do not think it makes sense to change or fix these methods as in your PR. You can have a look at this paper which solves this issue: https://arxiv.org/abs/2402.12030 as well as this blog https://huggingface.co/spaces/HuggingFaceH4/on-policy-distillation

**Comment by sambhavnoobcoder** [comment]
Hi @kashif Thanks for the references! I've reviewed both the paper and the blog post -and i feel they actually align with our approach. You're absolutely right about token-level KL not working - different token IDs and vocab sizes make direct token-level alignment impossible. That's exactly why we moved away from it. My approach is sequence-level KL, not token-level:

- I don't align token-by-token between vocabularies
- Instead, both models evaluate the same text (student generates, teacher gets re-tokenized version)
- Each model computes its own sequence-level log-probability: Σ log P(token_i) / N_tokens
- We can compare these sequence-level scores via KL: student_log_prob - teacher_log_prob

This is actually consistent with the blog post you linked - it also uses sequence-level probability as the bridge: "we compute the probability assigned by each model to the same text sequence" (from the On-Policy Distillation blog). The key insight is that text is the universal representation that bridges different tokenizer spaces. Both models can score the same text string, even if they tokenize it differently. What makes this different from token-level KL:

- Token-level: Compare P_student(token_i) vs P_teacher(token_i) → fails because token_i means different things
- Sequence-level: Compare P_student(text) vs P_teacher(text) → works because both evaluate the same text

If there's a fundamental issue with this approach, I'd love to understand it better. The referenced work seems to use similar sequence-level scoring. What specifically concerns you about this implementation?

**Comment by qgallouedec** [comment]
thanks for digging into this, and to kashif for the thoughtful back-and-forth. To make the disposition explicit: GKD and MiniLLM are formulated around token-level KL divergence between teacher and student logits, which intrinsically requires a shared vocabulary. That's not a bug we can patch around inside these trainers, it's the loss definition. TRL already ships an experimental cross-tokenizer trainer, `GOLDTrainer` in `trl/experimental/gold/`, which implements ULD (Universal Logit Distillation, https://arxiv.org/abs/2402.12030) and is designed precisely for the teacher/student-with-different-tokenizers case kashif linked. We'd much rather have improvements land there, or in a new trainer that follows the on-policy distillation recipe from the H4 blog, than retrofit GKD/MiniLLM. 

Closing this and #4561 on that basis, happy to chat in a fresh issue if you'd like to extend GOLD.
